cpu/stm32l1/stmclk.c

```diff
@@ -0,0 +1,242 @@
+ * Copyright (C) 2014 Freie Universität Berlin
+ *               2017 Fundacion Inria Chile
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32l1
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the cpu clock functions
+ *
+ * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Nick van IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ * @author      Víctor Ariño <victor.arino@zii.aero>
+ * @author      Oleg Artamonov <oleg@unwds.com>
+ * @author      Francisco Molina <francisco.molinas@inria.cl>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include "cpu.h"
+#include "stmclk.h"
+#include "periph_conf.h"
+
+/* See if we want to use the PLL */
+/* If HSE and HSI are both defined, then different defines for HSx DIV and MUL*/
+#if defined(CLOCK_PLL_DIV) || defined(CLOCK_PLL_MUL) || \
+    defined(CLOCK_PLL_DIV_HSE) || defined(CLOCK_PLL_MUL_HSE) || \
+    defined(CLOCK_PLL_DIV_HSI) || defined(CLOCK_PLL_MUL_HSI)
+#define CLOCK_USE_PLL              1
+#else
+#define CLOCK_USE_PLL              0
+#endif
+
+/* Make sure we have all needed information about the clock configuration */
+#if !defined(CLOCK_HSI) && !defined(CLOCK_HSE)
+#error "Please provide CLOCK_HSE or CLOCK_HSI in your board's perhip_conf.h"
+#endif
+#ifndef CLOCK_CORECLOCK
+#error "Please provide CLOCK_CORECLOCK in your board's periph_conf.h"
+#endif
+
+/* Check the source to be used for the PLL */
+#if defined(CLOCK_PLL_DIV_HSE) && defined(CLOCK_PLL_MUL_HSE) && \
+    defined(CLOCK_PLL_DIV_HSI) && defined(CLOCK_PLL_MUL_HSI)
+#define CLOCK_PLL_MULTI
+#elif defined(CLOCK_HSI)
+#define CLOCK_PLL_SOURCE           RCC_CFGR_PLLSRC_HSI
+#elif defined(CLOCK_HSE)
+#define CLOCK_PLL_SOURCE           RCC_CFGR_PLLSRC_HSE
+#endif
+
+/* Set core voltage according to core clock*/
+#if (CLOCK_CORECLOCK > 16000000U)
+#define CORE_VOLTAGE PWR_CR_VOS_0
+#elif (CLOCK_CORECLOCK > 8000000U)
+#define CORE_VOLTAGE PWR_CR_VOS_1
+#else
+#define CORE_VOLTAGE (PWR_CR_VOS_1 | PWR_CR_VOS_0)
+#endif
+
+static uint32_t tmpreg;
+static volatile uint32_t clock_source_rdy = 0;
+
+/**
+ * @brief Configure the clock system of the stm32l1
+ */
+void stmclk_init_sysclk(void)
+{
+    /* Reset the RCC clock configuration to the default reset state(for debug purpose) */
+    /* Set MSION bit */
+    RCC->CR |= RCC_CR_MSION;
+    /* Reset SW, HPRE, PPRE1, PPRE2, MCOSEL and MCOPRE bits */
+    RCC->CFGR &= ~(RCC_CFGR_HPRE | RCC_CFGR_PPRE1 | RCC_CFGR_PPRE2);
+    RCC->CFGR &= ~(RCC_CFGR_MCOSEL | RCC_CFGR_MCOPRE);
+    RCC->CFGR &= ~(RCC_CFGR_SW);
+    /* Reset PLL bits */
+    RCC->CFGR &= ~(RCC_CFGR_PLLSRC | RCC_CFGR_PLLDIV | RCC_CFGR_PLLMUL);
+    /* Reset HSION, HSEON, CSSON and PLLON bits */
+    RCC->CR &= ~(RCC_CR_HSION | RCC_CR_HSEON | RCC_CR_HSEBYP | RCC_CR_CSSON | RCC_CR_PLLON);
+    /* Clear all interrupts */
+    RCC->CIR = 0x0;
```

Comment:
Maybe this whole initialisation can follow a bit the one from L4, which takes into account the previous state of the interrupts.

Reply:
There's something I don't like about this approach. One ends up having multiple useless calls to irq_disable; irq_restore. This function in only used in cpu_init (are there any isr that should be restored here?), and in pm. And in pm ir_restore, irq_disable already has to be called around this function. Anyway, I saw this comment in th l4 code: `/* disable any interrupts. Global interrupts could be enabled if this is
     * called from some kind of bootloader...  */` . I don't now much about that case so if that is a concern I will include it, I leave it up to you @kYc0o, just feels weired habing so many nested calls to ir_disable/restore.

Reply:
Well actually I'm working on this platform because I want to support the RIOT bootlaoder I'm working on, towards OTA updates. Thus, the case of a bootlaoder is very important since I had several problems with clocking when it was already initialised by the bootloader. I can actually test it soon, as I'm approaching an end to the support of this platform (if #8518 gets merged soon)

Reply:
OK!, will add then.